Force Function Rename

brightprogrammer opened the pull request:
New function to forcefully rename an analysis function. Multiple retries will be performed by adding a numeric suffix to function name until the number of retries max's out

 <!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

...

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

brightprogrammer committed:
Force Function Rename

New function to forcefully rename a function. Multiple retries will be
performed by adding a numeric suffix to function name until the number
of retries max's out

brightprogrammer commented:
About [canalysis.c:3336](https://github.com/rizinorg/rizin/blob/64a4bb491a7cedb6381b1d0b0f34974ad856e13d/librz/core/canalysis.c#L3336) TODO?

wargio commented on librz/arch/function.c:
```diff
@@ -258,7 +258,22 @@ RZ_API bool rz_analysis_function_relocate(RzAnalysisFunction *fcn, ut64 addr) {
+ *
+ * @sa rz_analysis_function_force_rename
+ * */
 RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *name) {
```
add RZ_NOTNULL

wargio commented on librz/arch/function.c:
```diff
@@ -282,6 +297,38 @@ RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *nam
+ *
+ * @sa rz_analysis_function_rename
+ * */
+RZ_API const char *rz_analysis_function_force_rename(RzAnalysisFunction *fcn, const char *name) {
```
same here. RZ_NOT_NULL.

wargio commented on librz/arch/function.c:
```diff
@@ -282,6 +297,38 @@ RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *nam
+RZ_API const char *rz_analysis_function_force_rename(RzAnalysisFunction *fcn, const char *name) {
+	rz_return_val_if_fail(fcn && name, NULL);
+
+	const char *final_name = strdup(name);
+	ut32 retries = 100;
+	ut32 suffix = 0;
+	while ((retries--) && !rz_analysis_function_rename(fcn, final_name)) {
+		if (final_name) {
+			free((void *)final_name);
+			final_name = NULL;
+		}
+
+		ut32 strsz = snprintf(NULL, 0, "%s_%u", name, suffix) + 1;
+		final_name = malloc(strsz);
+		snprintf((char *)final_name, strsz, "%s_%u", name, suffix++);
+	}
+	return retries ? fcn->name : NULL;
```
instead of doing retries, just do `name_offset` which is granted to be unique. also use `rz_strf` directly instead of calling snprintf.

wargio reviewed (changes requested):
See above

wargio commented:
> About [canalysis.c:3336](https://github.com/rizinorg/rizin/blob/64a4bb491a7cedb6381b1d0b0f34974ad856e13d/librz/core/canalysis.c#L3336) TODO?

ignore it.

wargio commented:
Add a test also for this.

brightprogrammer committed:
Improve Force Rename & Add Tests

brightprogrammer committed:
Add RZ_NONNULL to Function Declarations

brightprogrammer committed:
Add RZ_BORROW To Force Rename

wargio commented on librz/arch/function.c:
```diff
@@ -310,23 +310,18 @@ RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *nam
  * @sa rz_analysis_function_rename
  * */
-RZ_API const char *rz_analysis_function_force_rename(RzAnalysisFunction *fcn, const char *name) {
+RZ_API const char *rz_analysis_function_force_rename(RZ_NONNULL RzAnalysisFunction *fcn, RZ_NONNULL const char *name) {
```
```suggestion
RZ_API RZ_BORROW const char *rz_analysis_function_force_rename(RZ_NONNULL RzAnalysisFunction *fcn, RZ_NONNULL const char *name) {
```

brightprogrammer replied:
This is already done [here](https://github.com/rizinorg/rizin/pull/4703/commits/057500dc42364527c5db3116e2e729d57d5109b3#diff-79ead514aa1e847124523b7f25695cb7e436de2494b5b09d7533a0ae1b3332baR313)

wargio commented on librz/arch/function.c:
```diff
@@ -310,23 +310,18 @@ RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *nam
+		// {name}_{addr} is guaranteed to be unique
+		rz_strf(final_name, "%s_%llx", name, fcn->addr);
+		return rz_analysis_function_rename(fcn, final_name) ? fcn->name : NULL;
 	}
```
remove the `else` since it is not needed.
Also do not use fixed size buffer for this but instead use `rz_str_newf` (sorry my bad, i mixed up names when i asked the previous changes) and free it after calling `bool ok = rz_analysis_function_rename(...)`, then `return ok ? fcn->name : NULL;`

wargio replied:
also use `PFMT64x` instead of `llx`, for example: `rz_str_newf("%s_%" PFMT64x, name, fcn->addr);` because `llx` is not portable.

brightprogrammer committed:
Apply Suggested Edits & Improve Tests

The whole point of force rename is that simple rename op might fail, so
I added a test to make sure it actually fails when we need it to...
hence the need for existence of force rename

wargio commented on test/unit/test_analysis_function.c:
```diff
@@ -468,6 +468,31 @@ bool test_noreturn_functions_list() {
+}
+
+bool test_analysis_function_force_rename() {
+	RzAnalysis *analysis = rz_analysis_new();
```
free rz_analysis

wargio commented on test/unit/test_analysis_function.c:
```diff
@@ -468,6 +468,31 @@ bool test_noreturn_functions_list() {
 }
 
+bool test_analysis_function_rename() {
+	RzAnalysis *analysis = rz_analysis_new();
```
free rz_analysis

wargio commented on test/unit/test_analysis_function.c:
```diff
@@ -468,6 +468,31 @@ bool test_noreturn_functions_list() {
+	rz_analysis_create_function(analysis, "b", 0xbabe, RZ_ANALYSIS_FCN_TYPE_FCN);
+
+	// rename a to b, but b already exists, so we force rename
+	mu_assert_notnull(rz_analysis_function_force_rename(a, "b"), "function force rename");
```
maybe check the output name.

wargio reviewed (changes requested):
there is no real need to call `RZ_FREE` since `free` was enough, but for me this is almost ok (besides the missing frees on rz_analysis and the missing check on the output name).
@thestr4ng3r

brightprogrammer committed:
Free Variables & More Checks Added To Tests

wargio commented on test/unit/test_analysis_function.c:
```diff
@@ -490,6 +495,11 @@ bool test_analysis_function_force_rename() {
 	// rename a to b, but b already exists, so we force rename
 	mu_assert_notnull(rz_analysis_function_force_rename(a, "b"), "function force rename");
+	// check whether the name begins with originally provided one
+	mu_assert_true(rz_str_startswith(a->name, "b"), "function renamed incorrectly");
```
i'm not yet ok with this test.
I think it should be two functions defined `a` and `b` (at different addresses), then i would force rename `a` to be `b` and check that is `b_xxxx` (since you know the address, it should be easy to guess the name).

wargio replied:
also check the return value matches with what is in  `a->name`

brightprogrammer replied:
I first thought about doing it that way, but then it assumes many things about the internal implementation. I just want to assume the function gets renamed with a unique suffix. This allows the implementation to be changed in future without breaking the test

wargio replied:
no. the function is meant to return a certain result which we know being name_offset, we should test for this, not for anything else. also if in the future we change something we can always change the test.

brightprogrammer replied:
Ok, I'll make the change :-)

brightprogrammer committed:
Check Full Name Of Renamed Function

Instead of just checking whether the starting of function matches with
expected name, check the complete name to be exactly same.

wargio reviewed (approved):
LGTM (as long as is green)

XVilka commented on librz/arch/function.c:
```diff
@@ -258,7 +258,22 @@ RZ_API bool rz_analysis_function_relocate(RzAnalysisFunction *fcn, ut64 addr) {
 
-RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *name) {
+/**
+ * @b Rename given function.
```
Use `\p`, etc Doxygen syntax

brightprogrammer replied:
Isn't this doxygen syntax as well?
> Structural commands (like [all other commands](https://doxygen.nl/manual/commands.html#cmd_intro)) start with a backslash (\), or an at-sign (@) if you prefer Javadoc style, followed by a command name and one or more parameters.

XVilka replied:
Yes, but we use the `\p` one: https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#documentation

brightprogrammer replied:
👍 sorry for not checking that 😅 , I'll make the change

brightprogrammer replied:
I have a habit of using `@` everywhere :-)

XVilka commented on librz/arch/function.c:
```diff
@@ -282,6 +297,34 @@ RZ_API bool rz_analysis_function_rename(RzAnalysisFunction *fcn, const char *nam
+ * This will make sure a function with non-unique name gets renamed by
+ * adding a numeric suffix to it.
+ *
+ * @param fcn
```
Same

brightprogrammer committed:
Change Documentation Style
